go/observability/observability.go

```diff
@@ -0,0 +1,29 @@
+package observability
+
+import (
+	"log"
+
+	"contrib.go.opencensus.io/exporter/stackdriver"
+	"go.opencensus.io/plugin/ocgrpc"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/trace"
+)
+
+func ConfigureStackdriver(gcpProject string) {
+	if err := view.Register(ocgrpc.DefaultClientViews...); err != nil {
+		log.Fatalf("Failed to register ocgrpc client views: %v", err)
+	}
+
+	sd, err := stackdriver.NewExporter(stackdriver.Options{
+		ProjectID: gcpProject,
+	})
+	if err != nil {
+		log.Fatalf("Failed to create Stackdriver exporter: %v", err)
+	}
+	defer sd.Flush()
+	trace.RegisterExporter(sd)
+	sd.StartMetricsExporter()
+	defer sd.StopMetricsExporter()
```

Comment:
This runs when `ConfigureStackdriver()` returns. I don't think we want to stop it here.

If `stop` does cleanup, we can return `sd`, and do the cleanup in `main()`.

And if we do that, I would also return the error.
```go
func ConfigureStackdriver(gcpProject string) (*stackdriver.NewExporter, error) {
```

---

And and, if we do need the goroutine to call `Flush`, we can wrap the exporter and the goroutine controlling channel together in another struct

```go
type exportAndFriends struct {
  exporter *stackdriver.NewExporter
  done channel struct{} // This can be used to exit the flushing goroutine
}
```

Reply:
Correct. I changed the function to return `*stackdriver.Exporter` and the caller (`main()`) now does the `defer sd.Flush()` and `defer sd.StopMetricsExporter()`.

There doesn't seem to be any need to return the error, as if an error is returned it will occur and exit the program prior to the call to `StartMetricsExporter()`. This is inline with the OpenCensus example along with the docs on the library itself: https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/blob/1996040a78017ba96c8bd972de452a9a7df00a04/stackdriver.go#L417